rnd/autogpt_server/autogpt_server/data/block.py

```diff
@@ -0,0 +1,130 @@
+import inspect
+import json
+
+from abc import ABC, abstractmethod
+from prisma.models import AgentBlock
+from pydantic import BaseModel
+from typing import ClassVar
+
+
+class Block(ABC, BaseModel):
+    @property
+    @abstractmethod
+    def input_schema(self) -> dict[str, str]:
+        """
+        The schema for the block input data. The keys are the names of the input,
+        and the values are the types of the input.
+        Example:
+        {
+            "system_prompt": "str",
+            "user_prompt": "str",
+            "max_tokens": "int",
+        }
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def output_schema(self) -> dict[str, str]:
+        """
+        The schema for the block possible output. The keys are the names of the output,
+        and the values are the types of the output.
+        Example:
+        {
+            "on_completion": "str",
+            "on_failure": "str",
+        }
+        """
+        pass
+
+    @abstractmethod
+    def run(self, input_data: dict[str, str]) -> (str, str):
```

Comment:
Type concern again, this means everything is passed as `str` and forces `Block` to validate or just fail, same with return value.

Reply:
the raw data will always passed as string since we won't store binary. We can provide a helper function like `get_input(name: str) -> Any` and it will parse it based on the type for convenience. Similar to output, we can json-stringify everything.

Reply:
> We can also add common validation layer on block (..)

That's probably a good idea, let's just avoid pushing validation as responsibility of block creators, it should be done automatically.

Reply:
I've added this on `execute` function within this class.